diff --git a/components/esp_hid/src/bt_hidh.c b/components/esp_hid/src/bt_hidh.c
index 6b33beee5e..53a99af52d 100644
--- a/components/esp_hid/src/bt_hidh.c
+++ b/components/esp_hid/src/bt_hidh.c
@@ -668,7 +668,7 @@ static void esp_hh_cb(esp_hidh_cb_event_t event, esp_hidh_cb_param_t *param)
                 if (param->data_ind.len == 9 && *(param->data_ind.data) == 1) {
                     has_report_id = true;
                     _usage = ESP_HID_USAGE_KEYBOARD;
-                } else if (param->data_ind.len == 4 && *(param->data_ind.data) == 2) {
+                } else if (param->data_ind.len >= 4 && *(param->data_ind.data) == 2) {
                     has_report_id = true;
                     _usage = ESP_HID_USAGE_MOUSE;
                 } else {
